polyfills.js

```diff
@@ -80,38 +81,120 @@ function patch (fs) {
     fs.lchownSync = function () {}
   }
 
-  // on Windows, A/V software can lock the directory, causing this
-  // to fail with an EACCES or EPERM if the directory contains newly
-  // created files.  Try again on failure, for up to 60 seconds.
-
-  // Set the timeout this long because some Windows Anti-Virus, such as Parity
-  // bit9, may lock files for up to a minute, causing npm package install
-  // failures. Also, take care to yield the scheduler. Windows scheduling gives
-  // CPU to a busy looping process, which can cause the program causing the lock
-  // contention to be starved of CPU by node, so the contention doesn't resolve.
+  // fs.rename and fs.renameSync uses MoveFileEx function on Windows.
+  // MoveFileEx is not atomic and honors Windows sharing modes, compared to
+  // os/syscall.rename used by Linux and OSX that are atomic and does not
+  // care if the file or directory is locked.
+  //
+  // This means that whenever a file or parent directory is locked (in use)
+  // on Windows the rename might fail with EACCS or EPERM errors depending
+  // the sharing mode set on the file and/or directory.
+  // 
+  // These win32-only overrides try to normalize fs.rename/renameSync
+  // behavior so it's more in line with how it works on Linux and OSX.
+  // It does this by retrying a failed rename for up to 5 seconds (or
+  // value of GRACEFUL_FS_WIN32_MAX_BACKOFF) until actually failing.
   if (platform === "win32") {
     fs.rename = (function (fs$rename) { return function (from, to, cb) {
+      try {
+        var stat = fs.statSync(to)
+        if (!stat) return
+        if (stat.isDirectory()) {
+          fs.rmdirSync(to)
```

Comment:
If I'm reading this correctly, it's synchronously unlinking the target prior to even attempting to rename.

That's... not how it works on Unix systems.

```
$ mkdir dir

$ touch file

$ node -e 'require("fs").renameSync("dir", "file")'
internal/fs/utils.js:259
    throw err;
    ^

Error: ENOTDIR: not a directory, rename 'dir' -> 'file'
    at Object.renameSync (fs.js:756:3)
    at [eval]:1:15
    at Script.runInThisContext (vm.js:131:20)
    at Object.runInThisContext (vm.js:297:38)
    at Object.<anonymous> ([eval]-wrapper:10:26)
    at Module._compile (internal/modules/cjs/loader.js:1176:30)
    at evalScript (internal/process/execution.js:94:25)
    at internal/main/eval_string.js:23:3 {
  errno: -20,
  syscall: 'rename',
  code: 'ENOTDIR',
  path: 'dir',
  dest: 'file'
}

$ node -e 'require("fs").renameSync("file", "dir")'
internal/fs/utils.js:259
    throw err;
    ^

Error: EISDIR: illegal operation on a directory, rename 'file' -> 'dir'
    at Object.renameSync (fs.js:756:3)
    at [eval]:1:15
    at Script.runInThisContext (vm.js:131:20)
    at Object.runInThisContext (vm.js:297:38)
    at Object.<anonymous> ([eval]-wrapper:10:26)
    at Module._compile (internal/modules/cjs/loader.js:1176:30)
    at evalScript (internal/process/execution.js:94:25)
    at internal/main/eval_string.js:23:3 {
  errno: -21,
  syscall: 'rename',
  code: 'EISDIR',
  path: 'file',
  dest: 'dir'
}
```